File: String/Python/Min Chars to Add for Palindrome.py
class Solution:
    def computeLPS(self,pat):
        n = len(pat)
        lps = [0] * n
        len_ = 0
        i = 1
        
        while i < n:
            if pat[i] == pat[len_]:
                len_ += 1
                lps[i] = len_
                i += 1
            else:
                if len_ != 0:
                    len_ = lps[len_ - 1]
                else:
                    lps[i] = 0
                    i += 1
        return lps 
        
        
    def minChar(self, s):
        # code here
        n = len(s)
        rev = s[::-1]
        s = s + "$" + rev
        lps = self.computeLPS(s)
        return n - lps[-1]

File: String/C++/Min Chars to Add for Palindrome.cpp
class Solution {
  public:
  
    vector<int> computeLPS(string &pat) {
        int n = pat.length();
        vector<int> lps(n);
        lps[0] = 0;
        int len_ = 0;
        int  i =1;
        while (i < n) {
            if (pat[i] == pat[len_]) {
                len_ += 1;
                lps[i] = len_;
                i++;
            }
            else {
                if (len_ != 0) {
                    len_ = lps[len_ - 1];
                }
                else {
                    lps[i] = 0;
                    i++;
                }
            }
        }
        return lps;
    }
    
    int minChar(string &s) {
        // code here
        int n = s.length();
        string rev = s;
        reverse(rev.begin(), rev.end());
        s = s + "\$" + rev;
        vector<int> lps = computeLPS(s);
        return n - lps.back();
    }
};
